app/models/compose_validator.rb

```diff
@@ -100,16 +100,16 @@ def validate_expose_format(value, path, errors)
       end
 
       def validate_service_name(value, path, errors)
-        regex = /^[a-zA-Z0-9]*$/
-        regex_with_build = /^[a-z0-9]*$/
+        regex = /^[a-zA-Z0-9]+(?:[._-][a-zA-Z0-9]+)*$/
+        regex_with_build = /^[a-z0-9]+(?:[._-][a-z0-9]+)*$/
         service_name = path[0]
         if value.include?('build')
           unless service_name =~ regex_with_build
-            errors << Kwalify::ValidationError.new('Invalid service name. Valid characters are [a-z0-9]', path)
+            errors << Kwalify::ValidationError.new('Invalid service name. Valid characters are [a-z0-9._-] but not starting with [.-_]', path)
           end
         else
           unless service_name =~ regex
-            errors << Kwalify::ValidationError.new('Invalid service name. Valid characters are [a-zA-Z0-9]', path)
+            errors << Kwalify::ValidationError.new('Invalid service name. Valid characters are [a-zA-Z0-9._-] but not starting with [.-_]', path)
```

Comment:
Is this true (that you can't begin with one of these characters)? They don't really mention it here: https://github.com/docker/compose/blob/fb4c9fbb8ba2797284296601eb904dd24c6a7b47/compose/config/validation.py#L27, but I assume you actually tried it...